system/jlib/jptree.cpp

```diff
@@ -8711,15 +8711,45 @@ class CConfigUpdater : public CInterface
     CriticalSection notifyFuncCS;
     unsigned notifyFuncId = 0;
     std::unordered_map<unsigned, ConfigUpdateFunc> notifyConfigUpdates;
+    std::vector<unsigned> pendingInitializeFuncIds;
 
 public:
-    CConfigUpdater(const char *_absoluteConfigFilename, IPropertyTree *_componentDefault, const char * * argv, const char * _componentTag, const char * _envPrefix, const char *_legacyFilename, IPropertyTree * (_mapper)(IPropertyTree *), const char *_altNameAttribute)
-        : absoluteConfigFilename(_absoluteConfigFilename), componentDefault(_componentDefault), componentTag(_componentTag), envPrefix(_envPrefix), legacyFilename(_legacyFilename), mapper(_mapper), altNameAttribute(_altNameAttribute)
+    CConfigUpdater()
     {
+    }
+    bool isInitialized() const
+    {
+        return args.ordinality(); // NB: null terminated, so always >=1 if initialized
+    }
+    void init(const char *_absoluteConfigFilename, IPropertyTree *_componentDefault, const char * * argv, const char * _componentTag, const char * _envPrefix, const char *_legacyFilename, IPropertyTree * (_mapper)(IPropertyTree *), const char *_altNameAttribute)
+    {
+        dbgassertex(!isInitialized());
+        absoluteConfigFilename.set(_absoluteConfigFilename);
+        componentDefault.set(_componentDefault);
+        componentTag.set(_componentTag);
+        envPrefix.set(_envPrefix);
+        legacyFilename.set(_legacyFilename);
+        mapper = _mapper;
+        altNameAttribute.set(_altNameAttribute);
         while (const char *arg = *argv++)
             args.append(arg);
         args.append(nullptr);
 
+        while (pendingInitializeFuncIds.size())
+        {
+            unsigned notifyFuncId = pendingInitializeFuncIds.back();
+            pendingInitializeFuncIds.pop_back();
+            ConfigUpdateFunc notifyFunc = notifyConfigUpdates[notifyFuncId];
+            notifyFunc(getComponentConfigSP(), getGlobalConfigSP());
```

Comment:
trivial: more efficient to evaluate these once outside the loop?  Fine as it is.

Reply:
I'll change it.